File: forecasto-server/src/forecasto/schemas/cashflow.py
"""Cashflow schemas."""

from __future__ import annotations


from datetime import date
from decimal import Decimal
from typing import Annotated

from pydantic import BaseModel, PlainSerializer


# Pydantic v2: Decimal fields serialized as float in JSON responses.
# The old DecimalAsFloat model_serializer approach was broken because
# Pydantic v2's default JSON handler converts Decimal → str before the
# custom serializer runs, so isinstance(obj, Decimal) never matched.
DecimalFloat = Annotated[Decimal, PlainSerializer(lambda v: float(v), return_type=float)]


class CashflowRequest(BaseModel):
    """Cashflow request parameters."""

    from_date: date
    to_date: date
    areas: list[str] | None = None
    stages: list[str] | None = None
    area_stage: list[str] | None = None  # format: "area:stage", e.g. "actual:0", "orders:1"
    bank_account_id: str | None = None
    group_by: str = "day"  # day, week, month
    sign_filter: str | None = None  # in, out, all

class CashflowRecordSummary(BaseModel):
    """Summary of a record in cashflow."""

    id: str
    reference: str
    amount: DecimalFloat
    area: str

class AccountCashflowEntry(BaseModel):
    """Cashflow data for a single bank account within a period."""

    inflows: DecimalFloat
    outflows: DecimalFloat
    running_balance: DecimalFloat

class CashflowEntry(BaseModel):
    """Single cashflow entry."""

    date: date
    inflows: DecimalFloat
    outflows: DecimalFloat
    net: DecimalFloat
    running_balance: DecimalFloat
    balance_snapshot: DecimalFloat | None = None  # set when a BankAccountBalance reset occurred on this date
    records: list[CashflowRecordSummary] | None = None
    by_account: dict[str, AccountCashflowEntry] | None = None

class AccountBalance(BaseModel):
    """Balance for a single account."""

    name: str
    balance: DecimalFloat
    credit_limit: DecimalFloat

class InitialBalance(BaseModel):
    """Initial balance information."""

    date: date
    total: DecimalFloat
    by_account: dict[str, AccountBalance]

class BalancePoint(BaseModel):
    """A point with date and balance."""

    date: date
    amount: DecimalFloat

class CashflowSummary(BaseModel):
    """Summary of cashflow calculation."""

    total_inflows: DecimalFloat
    total_outflows: DecimalFloat
    net_cashflow: DecimalFloat
    final_balance: DecimalFloat
    min_balance: BalancePoint
    max_balance: BalancePoint
    credit_limit_breaches: list[BalancePoint]

class CashflowResponse(BaseModel):
    """Cashflow response."""

    success: bool = True
    parameters: CashflowRequest
    initial_balance: InitialBalance
    cashflow: list[CashflowEntry]
    summary: CashflowSummary

class ConsolidatedCashflowRequest(BaseModel):
    """Consolidated cashflow request for multiple workspaces."""

    workspace_ids: list[str]
    from_date: date
    to_date: date
    group_by: str = "day"


# ── VAT Simulation schemas ─────────────────────────────────────────

class CashflowVatEntry(BaseModel):
    """Single VAT payment entry in cashflow simulation."""

    date: date
    period: str  # "2026-03" or "2026-Q1"
    area: str
    iva_debito: DecimalFloat
    iva_credito: DecimalFloat
    credit_carried: DecimalFloat
    net: DecimalFloat  # positive = da versare (uscita)


class CashflowVatSeries(BaseModel):
    """VAT simulation series for one P.IVA."""

    vat_registry_id: str
    vat_number: str
    name: str
    bank_account_id: str | None = None
    entries: list[CashflowVatEntry]
    total_debito: DecimalFloat
    total_credito: DecimalFloat
    total_net: DecimalFloat


class CashflowVatResponse(BaseModel):
    """Response with VAT simulation series."""

    series: list[CashflowVatSeries]


# ── Withholding Tax (Ritenuta d'Acconto) Simulation schemas ──────

class CashflowWithholdingEntry(BaseModel):
    """Single withholding payment entry."""

    date: date
    period: str  # "YYYY-MM"
    area: str
    amount: DecimalFloat  # positive = da versare (uscita)


class CashflowWithholdingResponse(BaseModel):
    """Response with withholding simulation entries."""

    entries: list[CashflowWithholdingEntry]
    total: DecimalFloat

File: forecasto-client-web/src/types/cashflow.ts
import type { Area } from './record'

export interface CashflowParams {
  from_date: string
  to_date: string
  areas: Area[]
  group_by: 'day' | 'week' | 'month'
  bank_account_id?: string
  project_code?: string
}

export interface AccountCashflowEntry {
  inflows: number
  outflows: number
  running_balance: number
}

export interface CashflowEntry {
  date: string
  inflows: number
  outflows: number
  net: number
  running_balance: number
  balance_snapshot?: number | null  // declared bank balance on this date (triggers reset)
  by_area: {
    budget?: { inflows: number; outflows: number }
    prospect?: { inflows: number; outflows: number }
    orders?: { inflows: number; outflows: number }
    actual?: { inflows: number; outflows: number }
  }
  by_account?: Record<string, AccountCashflowEntry>
}

export interface AccountBalance {
  name: string
  balance: number
  credit_limit: number
}

export interface InitialBalance {
  date: string
  total: number
  by_account: Record<string, AccountBalance>
}

export interface BalancePoint {
  date: string
  amount: number
}

export interface CashflowSummary {
  total_inflows: number
  total_outflows: number
  net_cashflow: number
  final_balance: number
  min_balance: BalancePoint
  max_balance: BalancePoint
  credit_limit_breaches: BalancePoint[]
}

export interface CashflowResponse {
  success: boolean
  parameters: CashflowParams
  cashflow: CashflowEntry[]
  summary: CashflowSummary
  initial_balance: InitialBalance
}

export interface BankAccount {
  id: string
  owner_id: string | null
  name: string
  bank_name: string | null
  description: string | null
  currency: string
  credit_limit: number
  is_active: boolean
  settings: Record<string, unknown>
  created_at: string
  updated_at: string
}

export interface BankAccountBalance {
  id: string
  bank_account_id: string
  balance_date: string
  balance: number
  source: string
  recorded_at: string
  note: string | null
}

export interface BankAccountBalanceCreate {
  balance_date: string
  balance: number
  source?: string
  note?: string
}

export interface BankAccountCreate {
  name: string
  bank_name?: string
  description?: string
  currency?: string
  credit_limit?: number
  settings?: Record<string, unknown>
}

export interface BankAccountUpdate {
  name?: string
  bank_name?: string
  description?: string
  currency?: string
  credit_limit?: number
  is_active?: boolean
  settings?: Record<string, unknown>
}

